<div style="background-color: #0000FF; color: white; padding: 10px; border-radius: 5px;">
  <h1>Team Info</h1>
</div>

## Team Name: Turbo Adventure
## Team Members: Akshay, Louis, Reno, Gabrijel
## Kaggle Usernames: 


<div style="background-color: #0000FF; color: white; padding: 10px; border-radius: 5px;">
  <h1>1 Feature Engineering</h1>
</div>

## 1.1 Imports

In [1]:
import os
import librosa
import numpy as np
import pandas as pd
from librosa.feature import rhythm  
from IPython.display import display
from sklearn.cluster import DBSCAN
import matplotlib.pyplot as plt

## 1.2 Defining functions

In [2]:
# ChatGPT, 2024, Prompt 1: "welke geluidsfeatures zijn er voor clustering van geluidsfragmenten"
# ChatGPT, 2024, Prompt 2: "geef per feature aan welke library ik daarvoor kan gebruiken"
# ChatGPT, 2024, Prompt 3: "laat voorbeeld code zien hoe deze features worden toegepast op een geluidsfragment, en vervolgens in een df wordt gezet"
# ChatGPT, 2024, Prompt 3: "hoe pas ik oop toe op deze code"
# Link: https://chatgpt.com/share/677ee38d-fb54-8001-a50a-9856d52e22c9

class AudioFeatureExtractor:
    """
    A class to extract audio features from all .wav files in a given folder.
    """

    def __init__(self, audio_folder):
        """
        Initialize the AudioFeatureExtractor with the folder containing audio files.
        """
        self.audio_folder = audio_folder
        self.features_list = []

    def process_audio_files(self):
        """
        Iterate through all .wav files in the folder and extract features.
        """
        for file_name in os.listdir(self.audio_folder):
            if file_name.endswith(".wav"):  # Only process .wav files
                file_path = os.path.join(self.audio_folder, file_name)
                self._extract_features(file_name, file_path)

    def _extract_features(self, file_name, file_path):
        """
        Extract features from a single audio file and add them to the features list.
        """
        try:
            # Load the audio file
            y, sr = librosa.load(file_path, sr=None)

            # Compute features
            features = {
                "File Name": file_name,
                "Zero Crossing Rate": np.mean(librosa.feature.zero_crossing_rate(y)[0]),
                "RMS Amplitude": np.mean(librosa.feature.rms(y=y)[0]),
                "Spectral Centroid": np.mean(librosa.feature.spectral_centroid(y=y, sr=sr)[0]),
                "Spectral Bandwidth": np.mean(librosa.feature.spectral_bandwidth(y=y, sr=sr)[0]),
                "Spectral Roll-Off": np.mean(librosa.feature.spectral_rolloff(y=y, sr=sr)[0]),
                "Spectral Flatness": np.mean(librosa.feature.spectral_flatness(y=y)[0]),
                "MFCC1": np.mean(librosa.feature.mfcc(y=y, sr=sr, n_mfcc=13)[0]),
                "MFCC2": np.mean(librosa.feature.mfcc(y=y, sr=sr, n_mfcc=13)[1]),
                "Chroma Features": np.mean(librosa.feature.chroma_cqt(y=y, sr=sr)),
                "Tempo (BPM)": rhythm.tempo(y=y, sr=sr)[0],
            }

            # Append features to the list
            self.features_list.append(features)

        except Exception as e:
            print(f"Error processing {file_name}: {e}")

    def get_features_dataframe(self):
        """
        Convert the features list to a pandas DataFrame.
        """
        return pd.DataFrame(self.features_list)

## 1.3 Extract Features & Create Dataframe

In [3]:
# ChatGPT, 2024, Prompt 1: "welke geluidsfeatures zijn er voor clustering van geluidsfragmenten"
# ChatGPT, 2024, Prompt 2: "geef per feature aan welke library ik daarvoor kan gebruiken"
# ChatGPT, 2024, Prompt 3: "laat voorbeeld code zien hoe deze features worden toegepast op een geluidsfragment, en vervolgens in een df wordt gezet"
# ChatGPT, 2024, Prompt 3: "hoe pas ik oop toe op deze code"
# Link: https://chatgpt.com/share/677ee38d-fb54-8001-a50a-9856d52e22c9

# Define the folder containing audio files
audio_folder_path = r"muziek-genre-clustering-24-25\unlabeled"

# Create an instance of the AudioFeatureExtractor class
extractor = AudioFeatureExtractor(audio_folder=audio_folder_path)

# Process the audio files
extractor.process_audio_files()

# Get the features as a DataFrame
audio_features_df = extractor.get_features_dataframe()

# Display the DataFrame
display(audio_features_df)

File Name  Zero Crossing Rate  RMS Amplitude  Spectral Centroid  \
0    m00003.wav            0.120116       0.136221        2254.606986   
1    m00012.wav            0.178108       0.217475        2908.300131   
2    m00013.wav            0.123423       0.029083        1952.607933   
3    m00043.wav            0.126073       0.100411        2384.634958   
4    m00044.wav            0.052088       0.018833         790.507005   
..          ...                 ...            ...                ...   
100  m00971.wav            0.091590       0.038885        1525.791709   
101  m00973.wav            0.161797       0.259782        3669.810467   
102  m00988.wav            0.167480       0.240724        3941.417697   
103  m00991.wav            0.044217       0.017138         983.971719   
104  m00995.wav            0.071629       0.066664        1253.032399   

     Spectral Bandwidth  Spectral Roll-Off  Spectral Flatness       MFCC1  \
0           2071.028440        4381.532206           0.009073  -82.511116   
1           2286.252592        5405.816551           0.026373   -1.913298   
2           1865.996047        3779.909832           0.002211 -287.603699   
3           2072.759900        4584.132502           0.008830 -120.147491   
4            900.409298        1159.412273           0.000153 -437.604401   
..                  ...                ...                ...         ...   
100         1588.217850        2733.184659           0.006208 -246.158646   
101         3295.063699        7876.372739           0.088763   -8.573005   
102         3361.641125        8260.423162           0.089354  -56.153061   
103         1404.408040        1853.129475           0.000302 -406.846680   
104         1371.258170        2292.769869           0.000566 -223.187332   

          MFCC2  Chroma Features  Tempo (BPM)  
0     97.341721         0.571940   135.999178  
1     72.686157         0.533881   103.359375  
2    101.758171         0.385155    95.703125  
3     91.317215         0.557109   135.999178  
4    170.931534         0.332215   107.666016  
..          ...              ...          ...  
100  149.428665         0.414986   107.666016  
101   54.295788         0.475779   129.199219  
102   36.448502         0.480985   103.359375  
103  163.081345         0.235405   117.453835  
104  163.253967         0.344185   151.999081  

[105 rows x 11 columns]

## 1.4 Features Explanation

### Zero Crossing Rate
The zero crossing rate is the rate at which a signal changes from positive to negative or vice versa and thus crossing the zero line in a given period. (11)


### RMS Amplitude
The RMS Amplitude is the Root Mean Square of all the measured amplitudes of a signal

### Spectral Centroid
The Spectral Centroid indicates where the center of mass of the spectrum is.

### Spectral Bandwidth

---

### **1. Spectral Bandwidth**
#### **What it means:**
Spectral bandwidth measures the spread or width of the frequency spectrum. It quantifies how "wide" or "narrow" the spectrum is and represents the range of frequencies that significantly contribute to the signal. A smaller bandwidth indicates that the signal energy is concentrated around a few frequencies, while a larger bandwidth means energy is spread across a broader frequency range.

#### **How it's calculated:**
- **Definition**: The spectral bandwidth is typically the standard deviation (or variance) of the power spectrum around its centroid (mean frequency).
- **Formula**:

$$
\text{Bandwidth} = \sqrt{\frac{\sum (f - f_c)^2 \cdot S(f)}{\sum S(f)}}
$$

Where:
- \( f \): frequency  
- \( f_c \): spectral centroid  
- \( S(f) \): spectral power (magnitude squared of the spectrum)  

#### **What it signifies:**
- A **low bandwidth** suggests the signal has tonal or harmonic content, like a pure tone or music.
- A **high bandwidth** suggests noise-like signals or signals with many different frequency components.

---

### Spectral Roll-Off
### **2. Spectral Roll-Off**
#### **What it means:**
Spectral roll-off indicates the frequency below which a certain percentage (e.g., 85%) of the total spectral energy is concentrated. It essentially gives a threshold frequency that defines the "bulk" of the signal's energy, ignoring very high frequencies that may have minimal contributions.

#### **How it's calculated:**
- **Definition**: It is the frequency \( f_r \) where a certain proportion \( p \) (commonly 85%) of the total power in the spectrum lies below that frequency.
- **Formula**:

$$
\sum_{f=0}^{f_r} S(f) \geq p \cdot \sum_{f=0}^{f_{\text{max}}} S(f)
$$

#### **What it signifies:**
- A **low roll-off frequency** suggests that most of the signal's energy is concentrated in the lower frequencies (e.g., a bass-heavy signal).
- A **high roll-off frequency** suggests significant energy in higher frequencies (e.g., signals with sharp transients or high harmonics).

---

### Spectral Flatness
### **3. Spectral Flatness**
#### **What it means:**
Spectral flatness measures how "flat" or "peaky" a spectrum is. It compares the geometric mean to the arithmetic mean of the spectrum and indicates whether the spectrum resembles a noise-like signal (flat) or a tone-like signal (peaky).

#### **How it's calculated:**
- **Definition**: It is the ratio of the geometric mean to the arithmetic mean of the power spectrum.
- **Formula**:

$$
\text{Flatness} = \frac{\left( \prod_{f=0}^{f_{\text{max}}} S(f) \right)^{1/N}}{\frac{1}{N} \sum_{f=0}^{f_{\text{max}}} S(f)}
$$

Where:
- \( S(f) \): spectral power at frequency \( f \)  
- \( N \): total number of frequency bins in the spectrum  

#### **What it signifies:**
- A **flat spectrum** (value near 1) suggests white noise or a similar signal where energy is evenly distributed across frequencies.
- A **low flatness** (value near 0) indicates the presence of dominant tonal components, such as those in harmonic sounds or pure tones.

---

### **Summary of Interpretations**:

| **Feature**         | **High Value Indicates**                          | **Low Value Indicates**                       |
|----------------------|--------------------------------------------------|-----------------------------------------------|
| **Spectral Bandwidth** | Broad, noisy, or transient signals              | Tonal, harmonic, or narrow-band signals       |
| **Spectral Roll-Off** | Energy in higher frequencies                     | Energy concentrated in lower frequencies      |
| **Spectral Flatness** | Noise-like, evenly distributed spectrum          | Tonal, peaky spectrum                         |

---

### MFCCs
### **X. MFCCs**
#### **What it means:**

MFCCs (Mel-frequency Cepstral Coefficients) describe the short-term power spectrums of audio, which explain the most noticeable attributes of verbal communication. Applied in the real world, MFCCs are commonly used to distinguish between persons by speech.

#### **How it's calculated:**

There is not only one single way to caculate MFCCs, but usually these steps are followed (GeeksforGeeks, 2024):

- Pre-emphasize the signal: Stabilizing the spectrum by raising the frequency
- Framing: Splitting the audio into more, but overlapping chunks
- Windowing: A window function is used to mitigate the edge effect of framing
- FFT (Fast Fourier Transform): The Fourier Transform is calculated by transforming the time domain frames to frequency domain frames in order to get the spectral representations
- Mel-filterbank: The Mel-filterbank seperates the spectral representations into Frequency Bands based on Mel-scale. Then filters extract the energy level and the more crucial frequencies get prioritized
- Logarithm: The logarithm is calculated to recieve a more accurate representation of the loudness
- DCT: After minimizing the duplicate coefficients, the Mel-frequency Cepstral Coefficients are finally the result

The rows of the MFCC matrix capture these features:
- MFCC1:
- MFCC2:

#### **What it signifies:**
- A **low MFCC1**
- A **high MFCC1**
- A **low MFCC2**
- A **high MFCC2**

### Chroma Features

Chroma features are a widely used representation in music analysis and processing, as they capture the harmonic content of a piece by summarizing the energy distribution across the twelve pitch classes (C, C#, D, etc.) within an octave. This abstraction allows for the comparison of tonal structures irrespective of the actual octave in which a note is played, making it particularly robust to changes in pitch and instrumentation. Müller (2015) highlights the effectiveness of chroma features in tasks such as chord recognition, genre classification, and music retrieval.

#### How do chroma features work?
The process of extracting chroma features involves the following steps:
1. **Frame segmentation:** The audio signal is divided into small, overlapping time frames.
2. **Frequency transformation:** Each frame undergoes a **Short-Time Fourier Transform (STFT)**, converting the audio from the time domain to the frequency domain.
3. **Pitch class aggregation:** For each frequency bin, the energy is mapped to its corresponding pitch class (e.g., C, C#, D), and the values are summed across all octaves. This results in a 12-dimensional chroma vector for each frame (Tzanetakis & Cook, 2002).

This process creates a "fingerprint" of the music's harmonic structure, enabling analysis that is invariant to timbral and octave shifts. For example, a melody played on a piano or a guitar will have a similar chroma representation if the notes are the same (Müller, 2015).

#### Example: Computing Chroma Features
Let’s take a sine wave with a frequency of 440 Hz (A4 note) sampled at 44.1 kHz:
1. Perform an STFT to compute the frequency spectrum of the audio signal.
2. Identify that 440 Hz corresponds to the pitch class "A."
3. Assign all energy to the pitch class "A" in the chroma vector, while all other pitch classes remain zero.

The resulting chroma vector might look like this:
`[0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0]`  
Here, the 8th position (A) has a value of 1, indicating that "A" is the dominant pitch class.

#### Applications and importance
- **Chord recognition:** Chroma features are widely used to detect harmonic progressions, such as identifying the chord sequence in a song (Müller, 2015).
- **Genre classification:** These features are particularly useful for distinguishing genres with complex harmonic structures, like jazz or classical music (Tzanetakis & Cook, 2002).
- **Music similarity and retrieval:** By focusing on tonal content, chroma features can help identify similar tracks or cover versions of songs (Wikipedia contributors, n.d.-a).

### Tempo (BPM)

Tempo refers to the speed or pace of a piece of music, measured in beats per minute (BPM). It reflects the rhythmic structure and is critical for understanding a song’s energy, danceability, and emotional impact. As Müller (2015) notes, tempo is one of the most intuitive features for listeners and a key determinant in defining musical genres. Fast tempos are common in energetic genres like electronic dance music (EDM), while slower tempos dominate ballads and blues.

#### How is tempo calculated?
The process of tempo extraction typically involves:
1. **Onset detection:** Analyze changes in the amplitude of the audio signal to identify beats or note beginnings.
2. **Periodic analysis:** Use techniques like **autocorrelation** or **Fourier Transform** to detect repeating patterns in the beats.
3. **Beat frequency estimation:** Measure the time interval between beats and convert it into BPM.

#### Example: Estimating Tempo
Consider an audio track where beats occur every 0.5 seconds:
1. Measure the time interval between beats: \( T = 0.5 \, \text{s} \).
2. Calculate the tempo using the formula:  
   $$\text{BPM} = \frac{60}{T} = \frac{60}{0.5} = 120 \, \text{BPM}.$$

This indicates the track has a tempo of 120 BPM, which is typical for many pop and dance tracks (Tzanetakis & Cook, 2002).

#### Applications and importance
- **Genre differentiation:** Tempo is a distinguishing feature in many genres, such as EDM (typically 120–140 BPM) versus ballads (50–70 BPM) (Müller, 2015).
- **Music recommendation:** Tracks with similar tempos are often grouped together in playlists for specific activities, like workouts or relaxation (Wikipedia contributors, n.d.-b).
- **Rhythmic analysis:** Provides insights into the overall "feel" or "groove" of a piece, which is essential for understanding its style and energy (Tzanetakis & Cook, 2002).

#

<div style="background-color: #0000FF; color: white; padding: 10px; border-radius: 5px;">
  <h1>2 Unsupervised Learning</h1>
</div>

## 2.1 K-Means

## 2.2 DBSCAN

DBSCAN (Density-Based Spatial Clustering of Applications with Noise) is a density-based clustering algorithm introduced by Ester et al. (1996). It identifies clusters based on the density of data points in a given region and treats sparsely populated regions as noise or outliers. Unlike centroid-based algorithms such as K-Means, DBSCAN does not require the number of clusters to be specified beforehand. Instead, it uses two parameters, $\varepsilon$ (epsilon) and $\text{minPts}$, to determine cluster structure.

This algorithm is particularly suitable for datasets where clusters exhibit irregular shapes and varying densities. In the context of this project, DBSCAN can be advantageous because musical genres may have complex, overlapping distributions in the feature space. Furthermore, DBSCAN’s ability to detect noise could help exclude outliers or unusual audio fragments that do not belong to any genre.

#### **Algorithmic Principles**

DBSCAN uses the following key concepts to identify clusters:

1. **Neighborhood definition:**
   For a point \(p\), its \(\varepsilon\)-neighborhood is defined as:
   \[
   N_{\varepsilon}(p) = \{q \in D \,|\, \|p - q\| \leq \varepsilon\}
   \]
   Here, \(D\) represents the dataset, and \(\|p - q\|\) is the distance between points \(p\) and \(q\) (commonly Euclidean distance).

2. **Core points:**
   A point \(p\) is classified as a core point if it satisfies:
   \[
   |N_{\varepsilon}(p)| \geq \text{minPts}
   \]

3. **Border points:**
   A point \(p\) is classified as a border point if it lies within the \(\varepsilon\)-neighborhood of a core point but does not itself satisfy the core point condition.

4. **Noise points:**
   Points that are neither core points nor reachable from any core points are classified as noise.

5. **Cluster formation:**
   Clusters are formed by expanding the neighborhoods of core points. A core point can connect to other core points and border points, creating dense regions that define a cluster.

#### **Algorithm Steps**

1. Begin with an unvisited point \(p\).
2. Determine if \(p\) is a core point:
   - If \(p\) is a core point, start a new cluster.
   - Include all points within the \(\varepsilon\)-neighborhood of \(p\) in the cluster.
   - Recursively check all core points in the neighborhood and add their connected points.
3. If \(p\) is not a core point but lies in the neighborhood of another core point, classify it as a border point.
4. If \(p\) is neither a core nor a border point, classify it as noise.
5. Repeat until all points in the dataset have been visited.

#### **Advantages and Limitations**

**Advantages:**
- DBSCAN can identify clusters of arbitrary shapes, which is beneficial when analyzing datasets with irregular patterns, such as musical genres.
- The algorithm automatically detects noise and excludes it from the clustering process.
- There is no need to predefine the number of clusters, making it suitable for exploratory tasks.

**Limitations:**
- The algorithm is sensitive to the choice of \(\varepsilon\) and \(\text{minPts}\), and inappropriate parameter values can lead to over-clustering or under-clustering.
- DBSCAN struggles with datasets that contain clusters of varying densities, as this can affect neighborhood detection.

#### **Application Example**

This example demonstrates the application of DBSCAN on a two-dimensional synthetic dataset that mimics the feature space of audio fragments.

In [ ]:
# Generate synthetic data
np.random.seed(42)
dense_cluster = np.random.randn(100, 2) * 0.5 + [2, 2]
sparse_cluster = np.random.randn(50, 2) * 0.8 + [-2, -2]
noise = np.random.uniform(-6, 6, (20, 2))
data = np.vstack([dense_cluster, sparse_cluster, noise])

# Apply DBSCAN
dbscan = DBSCAN(eps=0.8, min_samples=5)
labels = dbscan.fit_predict(data)

# Visualize the results
plt.figure(figsize=(8, 6))
unique_labels = set(labels)
for label in unique_labels:
    cluster_points = data[labels == label]
    if label == -1:
        plt.scatter(cluster_points[:, 0], cluster_points[:, 1], color='red', label='Noise')
    else:
        plt.scatter(cluster_points[:, 0], cluster_points[:, 1], label=f'Cluster {label}')
plt.title('DBSCAN Clustering')
plt.xlabel('Feature 1')
plt.ylabel('Feature 2')
plt.legend()
plt.show()

#### Conclusion

DBSCAN is a suitable algorithm for this project because of its ability to identify clusters of arbitrary shapes and handle noise effectively. Unlike centroid-based methods such as K-Means, DBSCAN does not assume spherical clusters and is well-suited for datasets with irregular patterns (Ester et al., 1996). This flexibility is particularly advantageous when clustering musical genres, as the feature space may exhibit overlapping or complex structures.

Moreover, DBSCAN's robustness against noise allows it to exclude outliers from the clustering process, making it effective for detecting unusual or out-of-distribution audio fragments (Pedregosa et al., 2011). However, the algorithm is sensitive to the choice of its parameters ($\varepsilon$ and $\text{minPts}$), which require careful tuning based on the dataset's density distribution.

Despite these challenges, DBSCAN remains a strong candidate for exploratory clustering in this project due to its adaptability and lack of reliance on predefined cluster counts. Future steps should include parameter optimization and comparison with other clustering methods to assess its performance comprehensively.

## 2.3 Gaussian Mixture Models (GMM)

GMM is a soft clustering model, that assigns each observation a probability of belonging to a specific cluster. In this algorithm each cluster is represented by a normal (gaussian) distribution with the mean and covariance matrix as metrics.

This model utilizes the Expectation-Maximization method, which is split into the following phases:

- Initialize phase: The mean, the covariance matrix and the mixing coefficients of every gaussian distribution (cluster) are assumed.

- Expectation phase: The observations are assigned to gaussian distributions (clusters) based on the calculated probability, that they belong to them.

- Maximization phase: The mean, the covariance matrix and the mixing coefficients of every gaussian distribution (cluster) are calculated with the assigned observations. 

Then the Expectation and Maximization phase are reiterated, until the model can not be notably improved anymore.

To calculate the probability, that an observation belongs to a cluster, the following probability density function (pdf) formula is used (What Is Gaussian Mixture Model | Deepchecks, 2023):

<br>

$$
\text{pdf}(x) = \sum_{k=1}^{K} \pi_k \cdot N(x|\mu_k, \Sigma_k)
$$

- $\pi_k$ = Mixing coefficient
- $\mu_k$ = Mean vector  
- $\Sigma_k$ = Covariance matrix
- $N(x|\mu_k, \Sigma_k)$ = Probability density function

<div style="background-color: #0000FF; color: white; padding: 10px; border-radius: 5px;">
  <h1>3 Findings and Conclusion</h1>
</div>

## 3.1

<div style="background-color: #0000FF; color: white; padding: 10px; border-radius: 5px;">
  <h1>4 Reference List</h1>
</div>

## 4.1

### Online Sources

* ChatGPT, 2024, Prompt 1: "welke geluidsfeatures zijn er voor clustering van geluidsfragmenten", (https://chatgpt.com/share/677ee38d-fb54-8001-a50a-9856d52e22c9)

* ChatGPT, 2024, Prompt 2: "geef per feature aan welke library ik daarvoor kan gebruiken", (https://chatgpt.com/share/677ee38d-fb54-8001-a50a-9856d52e22c9)

* ChatGPT, 2024, Prompt 3: "laat voorbeeld code zien hoe deze features worden toegepast op een geluidsfragment, en vervolgens in een df wordt gezet", (https://chatgpt.com/share/677ee38d-fb54-8001-a50a-9856d52e22c9)
* ChatGPT, 2024, Prompt 4: "hoe pas ik oop toe op deze code", (https://chatgpt.com/share/677ee38d-fb54-8001-a50a-9856d52e22c9)
* GeeksforGeeks. (2024, 26. Juni). Melfrequency Cepstral Coefficients (MFCC) for Speech Recognition. GeeksforGeeks. https://www.geeksforgeeks.org/mel-frequency-cepstral-coefficients-mfcc-for-speech-recognition/
* What is Gaussian Mixture Model | Deepchecks. (2023, 23. Januar). Deepchecks. https://www.deepchecks.com/glossary/gaussian-mixture-model/
- Wikipedia contributors. (n.d.-a). Chroma feature. In *Wikipedia*. Retrieved January 12, 2025, from  
   [https://en.wikipedia.org/wiki/Chroma_feature](https://en.wikipedia.org/wiki/Chroma_feature)

-  Wikipedia contributors. (n.d.-b). Tempo. In *Wikipedia*. Retrieved January 12, 2025, from  
   [https://en.wikipedia.org/wiki/Tempo](https://en.wikipedia.org/wiki/Tempo)


### Literature

- Ester, M., Kriegel, H.-P., Sander, J., & Xu, X. (1996). A Density-Based Algorithm for Discovering Clusters in Large Spatial Databases with Noise. *Proceedings of the 2nd International Conference on Knowledge Discovery and Data Mining (KDD)*, 226–231.  
  Retrieved from [https://www.aaai.org/Papers/KDD/1996/KDD96-037.pdf](https://www.aaai.org/Papers/KDD/1996/KDD96-037.pdf)

- Müller, M. (2015). *Fundamentals of Music Processing: Audio, Analysis, Algorithms, Applications*. Springer Verlag.  
   This comprehensive textbook covers key topics in music information retrieval, including chroma features and tempo analysis.

- Pedregosa, F., Varoquaux, G., Gramfort, A., Michel, V., Thirion, B., Grisel, O., ... & Duchesnay, É. (2011). Scikit-learn: Machine Learning in Python. *Journal of Machine Learning Research*, 12, 2825–2830.  
  Retrieved from [https://jmlr.csail.mit.edu/papers/v12/pedregosa11a.html](https://jmlr.csail.mit.edu/papers/v12/pedregosa11a.html)

-  Tzanetakis, G., & Cook, P. (2002). Musical genre classification of audio signals. *IEEE Transactions on Speech and Audio Processing*, 10(5), 293–302.  
   DOI: [10.1109/TSA.2002.800560](https://doi.org/10.1109/TSA.2002.800560)